content/blog/2019-11-13-llvm-profiling/index.md

```diff
@@ -0,0 +1,336 @@
+[Yi Jiang](http://www.cs.cornell.edu/~yijiang/) is a 2nd year Phd student in CS dept, Cornell. She is interested in designing compilers for Processing-In-Memory related architectures and also general compiler optimization. \n
+[Neil Adit](http://www.neiladit.com) is (guess what!) also a 2nd year PhD student at CSL, Cornell. He is interested in fun things in compiler research.
+
+"""
+extra.latex = true
+
++++
+
+
+
+## Goal
+
+The goal of the project is to collect run-time information by adding an LLVM pass. We are interested in three kinds of information: 
+
+1. Basic execution information including the number of LLVM instructions, basic blocks etc.
+2. Expensive instruction information like the number of memory operations, multiplications and branches since they are most likely to affect execution time.
+3. Function call information: the number of times each function is called.
+
+
+
+## Design Overview
+
+The LLVM architecture looks like this:
+
+<img src="llvm.jpg" style="width: 90%">
+
+
+
+IR represents intermediate language. We only need to add one pass that changes the existing IR to another IR (which includes the desired properties like profiling information), following instructions to add an LLVM pass in our instructor, [Adrian](<https://www.cs.cornell.edu/~asampson/>)'s [blog post](<https://www.cs.cornell.edu/~asampson/blog/llvm.html>). We also took help from the github repo [atomicCounter](<https://github.com/pranith/AtomicCounter/blob/master/AtomicCountPass/AtomicCount.cpp>) where it counts the number of all atomic instructions in the program atomically!
+
+We start from [ModulePass](<http://llvm.org/docs/WritingAnLLVMPass.html#the-modulepass-class>) and then access functions inside a Module, Basic Blocks per function and instructions per Basic Block to obtain all information. At the beginning and ending of each module, we also call custom method `initialize` to create global variables in IR  and `finalize` to print global variables. The structure of our atomic counting pass looks like this:
+
+```c++
+  struct SkeletonPass : public ModulePass {
+    static char ID;
+    SkeletonPass() : ModulePass(ID) {}
+      
+    virtual bool runOnModule(Module &M); //when there is a Module
+    virtual bool runOnFunction(Function &F, Module &M); //called by runOnModule
+    virtual bool runOnBasicBlock(BasicBlock &BB, Module &M); // called by runOnFunction
+      
+    bool initialize(Module &M); //create global variable
+    bool finialize(Module &M); //print global variable
+      
+    void createInstr(BasicBlock &bb, Constant *counter_ptr, int num);
+      
+    vector<string> atomicCounter; //keep global variable names for profiling. e.g. instr counter
+  };
+```
+
+
+
+## Implementation
+
+### Basic Execution Information
+
+We first create global variable `llvmInstrAtomicCounter` and `basicBlockAtomicCounter` with [GlobalVariable](<https://llvm.org/doxygen/classllvm_1_1GlobalVariable.html#a3ef813d6bda7e49e31cb6bf239c4e264>) constructor.
```

Comment:
Hmm; this seems to differ from the code listing above, which doesn't list these variables.

Reply:
That's in initialize function. We didn't post it in the blog.